qiskit/quantum_info/synthesis/qsd.py

```diff
@@ -0,0 +1,196 @@
+    """
+    Decomposes unitary matrix into one and two qubit gates using Quantum Shannon Decomposition.
+
+       ┌───┐               ┌───┐     ┌───┐     ┌───┐
+      ─┤   ├─       ───────┤ Rz├─────┤ Ry├─────┤ Rz├─────
+       │   │    ≃     ┌───┐└─┬─┘┌───┐└─┬─┘┌───┐└─┬─┘┌───┐
+     /─┤   ├─       /─┤   ├──□──┤   ├──□──┤   ├──□──┤   ├
+       └───┘          └───┘     └───┘     └───┘     └───┘
+
+    The number of CX gates generated with the decomposition without optimizations is,
+
+    .. math::
+
+        \frac{9}{16} 4^n - frac{3}{2} 2^n
+
+    If opt_a1=True, the CX count is further reduced by,
+
+    .. math::
+
+        \frac{1}{3} 4^{n - 2} - 1
+
+    Arguments:
+       mat (ndarray): unitary matrix to decompose
+       opt_a1 (bool): whether to try optimization A.1 from Shende. This should elliminate 1 cnot
+          per call. If True CZ gates are left in the output. If desired these can be further decomposed
+          to CX.
+       decomposer_1q (None or Object): optional 1Q decomposer. If None, uses
+          :class:`~qiskit.quantum_info.synthesis.one_qubit_decomposer.OneQubitEulerDecomser`
+       decomposer_2q (None or Object): optional 2Q decomposer. If NOne, uses
+          :class:`~qiskit.quantum_info.synthesis.two_qubit_decomposer.TwoQubitBasisDecomposer`
+          with CXGate.
+
+    Return:
+       QuantumCircuit: Decomposed quantum circuit.
+    """
+    dim = mat.shape[0]
+    nqubits = int(np.log2(dim))
+    if np.allclose(np.identity(dim), mat):
+        return QuantumCircuit(nqubits)
+    if dim == 2:
+        if decomposer_1q is None:
+            decomposer_1q = one_qubit_decompose.OneQubitEulerDecomposer()
+        circ = decomposer_1q(mat)
+    elif dim == 4:
+        if decomposer_2q is None:
+            decomposer_2q = two_qubit_decompose.TwoQubitBasisDecomposer(CXGate())
+        circ = decomposer_2q(mat)
+    else:
+        qr = QuantumRegister(nqubits)
+        circ = QuantumCircuit(qr)
+        dim_o2 = dim // 2
+        # perform cosine-sine decomposition
+        (u1, u2), vtheta, (v1h, v2h) = scipy.linalg.cossin(mat, separate=True, p=dim_o2, q=dim_o2)
+        # left circ
+        left_circ = demultiplex(v1h, v2h, opt_a1=opt_a1)
+        circ.append(left_circ.to_instruction(), qr)
+        # middle circ
+        if opt_a1:
+            nangles = len(vtheta)
+            half_size = nangles // 2
+            # get UCG in terms of CZ
+            circ_cz = _get_ucry_cz(nqubits, (2 * vtheta).tolist())
+            circ.append(circ_cz.to_instruction(), range(nqubits))
+            # merge final cz with right-side generic multiplexer
+            u2[:, half_size:] = np.negative(u2[:, half_size:])
+        else:
+            circ.ucry((2 * vtheta).tolist(), qr[:-1], qr[-1])
+        # right circ
+        right_circ = demultiplex(u1, u2, opt_a1=opt_a1)
+        circ.append(right_circ.to_instruction(), qr)
+
+    return circ
+
+
+def demultiplex(um0, um1, opt_a1=False):
```

Comment:
Originally I made it public because it seemed to express a generally useful decomposition. However, with the way it is currently written, referring to qs_decomposition, perhaps it could be hidden.

Reply:
3880ffecb